azurelinuxagent/ga/exthandlers.py

```diff
@@ -310,11 +311,74 @@ def handle_ext_handlers(self, etag=None):
                 logger.info("Extension handling is on hold")
                 return
 
+        wait_until = datetime.datetime.utcnow() + datetime.timedelta(minutes=DEFAULT_EXT_TIMEOUT_MINUTES)
+        prev_handler = None
+        ext_seq_required = self.is_ext_seq_required()
+
         self.ext_handlers.extHandlers.sort(key=operator.methodcaller('sort_key'))
         for ext_handler in self.ext_handlers.extHandlers:
-            # TODO: handle install in sequence, enable in parallel
+            if ext_seq_required:
```

Comment:
I would suggest switching the order of the calls to handle_ext_handler and wait_for_prev_handler_completion (call handle first and wait second) and just add a check to avoid waiting for the last extension. I think this express the intention of the code a little more clearly.

Reply:
Please don't use `for i in range(L)` in a Python 2 or Python 2/3 codebase such as this one. It creates a list of integers unnecessarily.

For this particular case, keeping track of `i` is unnecessary. The second to last `ext_handler` may also be a leaf node so the test needs to be against the maximum dependency level, not against its position in the list. `if dep_level >= 0 and dep_level < max_dep_level:` ought to be a sufficient test.

Reply:
agree